x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolver.java

```diff
@@ -222,64 +234,151 @@ private void filterResults(String javaRegex, GetAliasesResponse aliases, GetInde
-        client.admin().indices().getIndex(getIndexRequest, ActionListener.wrap(response -> {
-            ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = response.getMappings();
-
-            List<IndexResolution> resolutions;
-            if (mappings.size() > 0) {
-                resolutions = new ArrayList<>(mappings.size());
-                Pattern pattern = javaRegex != null ? Pattern.compile(javaRegex) : null;
-                for (ObjectObjectCursor<String, ImmutableOpenMap<String, MappingMetaData>> indexMappings : mappings) {
-                    String concreteIndex = indexMappings.key;
-                    if (pattern == null || pattern.matcher(concreteIndex).matches()) {
-                        resolutions.add(buildGetIndexResult(concreteIndex, concreteIndex, indexMappings.value));
+    public void resolveAsMergedMapping(String indexWildcard, String javaRegex, ActionListener<IndexResolution> listener) {
+        FieldCapabilitiesRequest fieldRequest = createFieldCapsRequest(indexWildcard);
+        client.fieldCaps(fieldRequest,
+                ActionListener.wrap(response -> listener.onResponse(mergedMapping(indexWildcard, response.get())), listener::onFailure));
+    }
+
+    static IndexResolution mergedMapping(String indexPattern, Map<String, Map<String, FieldCapabilities>> fieldCaps) {
+        if (fieldCaps == null || fieldCaps.isEmpty()) {
+            return IndexResolution.notFound(indexPattern);
+        }
+
+        StringBuilder errorMessage = new StringBuilder();
+
+        NavigableSet<Entry<String, Map<String, FieldCapabilities>>> sortedFields = new TreeSet<>(
+                // for some reason .reversed doesn't work (prolly due to inference)
+                Collections.reverseOrder(Comparator.comparing(Entry::getKey)));
+        sortedFields.addAll(fieldCaps.entrySet());
+
+        Map<String, EsField> hierarchicalMapping = new TreeMap<>();
+        Map<String, EsField> flattedMapping = new LinkedHashMap<>();
+        
+        // sort keys descending in order to easily detect multi-fields (a.b.c multi-field of a.b)
+        // without sorting, they can still be detected however without the emptyMap optimization
+        // (fields without multi-fields have no children)
+        for (Entry<String, Map<String, FieldCapabilities>> entry : sortedFields) {
+            String name = entry.getKey();
+            // skip internal fields
+            if (!name.startsWith("_")) {
+                Map<String, FieldCapabilities> types = entry.getValue();
+                // field is mapped differently across indices
+                if (types.size() > 1) {
+                    // build error message
+                    for (Entry<String, FieldCapabilities> type : types.entrySet()) {
+                        if (errorMessage.length() > 0) {
+                            errorMessage.append(", ");
+                        }
+                        errorMessage.append("[");
```

Comment:
Is this tested with multiple entries?

Reply:
If you're asking about the error message, yes in `IndexResolverTests` (e.g. `testMergeIncompatibleTypes`).